# <center>**TOÁN ỨNG DỤNG VÀ THỐNG KÊ**</center>
## <center>**Project 4 - Linear Regression**</center>

***
## **1. Thông tin sinh viên**
- Họ và tên: Nguyễn Thái Bảo
- Mã số sinh viên: 23120023
- Lớp: 23CTT1
- Email: 23120023@student.hcmus.edu.vn

***
## **2. Giải thuật cho bài toán Least squares data fitting**

### **Tổng quan về bài toán khớp dữ liệu (data fitting)**

Một trong những bài toán căn bản trong các ngành khoa học ứng dụng là khảo sát sự liên hệ giữa một đại lượng, ký hiệu y, theo một số đại lượng khác, ký hiệu $x_1. x_2, ..., x_n$. Tiếp cận đơn giản nhất là tìm một hàm số f: $R^n \rightarrow R$ sao cho y ≈ f(x).

Trong đó x = $(x_1, x_2, ..., x_n) \in R^n$ được gọi là vector các biến độc lập (independent variables), và $y \in R$ được gọi là biến phụ thuộc (response variable).


### **Xét bài toán đơn giản**

Cho 2 đại lượng x, y. Ta cần tìm hàm để biểu diễn y theo x.

Xét hàm $f(x) = \beta_0 + \beta_1x + \beta_2x^2 + ... + \beta_nx^n$

Hàm $f(x)$ tốt khi: $$\sum [y_i - (\beta_0 + \beta_1x + \beta_2x^2 + ... + \beta_nx^n)]^2 \rightarrow min$$

Ta có:
$$
\begin{aligned}
SSE &= \sum_{i=1}^{n} (\beta_0 + \beta_1x_i + \beta_2x_i^2 + ... + \beta_mx_i^m - y_i)^2 \rightarrow min \\
&= L(\beta) \\
&= ||A\beta - Y||^2 \rightarrow min
\end{aligned}
$$

Trong đó:

$$\beta = \begin{pmatrix}
\beta_0 \\
\beta_1 \\
\vdots \\
\beta_n
\end{pmatrix}, \quad

Y = \begin{pmatrix}
y_1 \\
y_2 \\
\vdots \\
y_n
\end{pmatrix}, \quad

A = \begin{pmatrix}
1       & x_1    & x_1^2    & \cdots    & x_1^n \\
1       & x_2    & x_2^2    &\cdots     & x_2^n \\
\vdots  & \vdots & \vdots   &\ddots     & \vdots \\
1       & x_n    & x_n^2    &\cdots     & x_n^n
\end{pmatrix}$$

Mặt khác, $L(x)$ đạt giá trị nhỏ nhất tại $\nabla L = 0$
$$\Leftrightarrow 2A^TAx - 2A^Tb = 0$$
$$\Leftrightarrow x = (A^TA)^{-1}(A^Tb)$$

Để $L(\beta)$ khớp tốt nhất cho dữ liệu:
$$\hat{\beta} = (A^TA)^{-1}(A^Ty)$$



### **Mô hình tuyến tính theo tham số**

Trong phần này, ta tập trung khảo sát mô hình tuyến tính theo tham số (linear in the parameters model) cho lớp hàm dự báo:

$$\hat{f}(x) = \theta_1 f_1(x) + \cdots + \theta_p f_p(x) \quad (1)$$

Trong đó:
- $f_1, f_2, ..., f_p$: $R^n \rightarrow R$ là các hàm cơ sở (basis functions) được xác định trước
- $\theta_1, \theta_2, ..., \theta_p$ là các tham số mô hình (model parameters) cần tìm sao cho hàm dự báo tương ứng là phù hợp (consistent) với dữ liệu thu thập được.



### **Đánh giá sự phù hợp của mô hình**

Để đánh giá sự phù hợp của mô hình (1) với dữ liệu, ta tính các giá trị dự báo của mô hình tương ứng với giá trị của các biến độc lập của dữ liệu:

$$\hat{y}^{(i)} = \hat{f}(x^{(i)}) = \hat{f}(x_1^{(i)}, ..., x_n^{(i)}) \quad \text{với} \quad i = 1, ..., N$$

và đánh giá các sai số dự báo (prediction error) hay phần dư (residual):

$$r^{(i)} = \hat{y}^{(i)} - y^{(i)} \quad \text{với} \quad i = 1, ..., N \quad (2)$$



### **Phương pháp bình phương nhỏ nhất (OLS)**

Phương pháp bình phương nhỏ nhất tìm cách xác định các tham số $\theta_i$ của mô hình (1) sao cho vector các phần dư:

$$r = (r^{(1)}, ..., r^{(N)}) \in \mathbb{R}^N \quad (3)$$

có chuẩn nhỏ nhất.

Từ (1) và (2), ta có:

$$r^{(i)} = \theta_1 f_1(x^{(i)}) + \cdots + \theta_p f_p(x^{(i)}) - y^{(i)}$$

$$= \begin{pmatrix} f_1(x^{(i)}) & \cdots & f_p(x^{(i)}) \end{pmatrix} 
\begin{pmatrix} \theta_1 \\ \vdots \\ \theta_p \end{pmatrix} - y^{(i)} \quad \text{với} \quad i = 1, ..., N$$

Bằng cách đặt:

$$A = \begin{pmatrix} 
f_1(x^{(1)}) & \cdots & f_p(x^{(1)}) \\
\vdots & \ddots & \vdots \\
f_1(x^{(N)}) & \cdots & f_p(x^{(N)})
\end{pmatrix} \in R^{N \times p}$$

$$\theta = \begin{pmatrix} \theta_1 \\ \vdots \\ \theta_p \end{pmatrix} \in R^p, \quad 
y = \begin{pmatrix} y^{(1)} \\ \vdots \\ y^{(N)} \end{pmatrix} \in R^N$$

Ta có $r = A\theta - y$. Do áp dụng kết quả nghiệm của bài toán bình phương tối tiểu, ta có:

$$\theta = (A^T A)^{-1} A^T y$$

### **Đánh giá mô hình**

Để đánh giá mô hình, ta có thể sử dụng các chỉ số:
- Mean Squared Error (MSE)
- Hệ số xác định ($R^2$)
- Chuẩn vector phần dư (Residual norms)

Qua các chỉ số này, ta có thể so sánh và chọn ra mô hình tốt nhất phù hợp với bài toán.

***
## **3. Các hàm tiện ích (Utils)**

In [1]:
def sqrt(x):
    """Tính căn bậc hai của một số"""
    return round(x ** 0.5, 5) if x >= 0 else None

In [2]:
# Các hàm sử dụng cho ma trận

def create_matrix(rows, cols, default_value = 0):
    """Tạo ma trận với rows hàng và cols cột"""
    return [[default_value for _ in range(cols)] for _ in range(rows)]

def create_identity_matrix(n):
    """Tạo ma trận đơn vị"""
    identity = create_matrix(n, n)
    for i in range(n):
        identity[i][i] = 1
    return identity

def matrix_copy(A):
    """Tạo bản sao của ma trận A"""
    return [row[:] for row in A]

def matrix_multiply(A, B):
    """Nhân hai ma trận A và B"""

    # Trường hợp B là vector
    if not isinstance(B[0], list):
        # Chuyển B thành ma trận cột
        B = [[b] for b in B]
    
    # Lấy kích thước của ma trận
    rows_A = len(A)
    cols_A = len(A[0])
    rows_B = len(B)
    cols_B = len(B[0]) if isinstance(B[0], list) else 1
    
    # Kiểm tra xem có thể nhân ma trận không
    if cols_A != rows_B:
        raise ValueError(f"Không thể nhân ma trận với kích thước không phù hợp: ({rows_A}x{cols_A}) và ({rows_B}x{cols_B})")
    
    # Khởi tạo ma trận kết quả C kích thước rows_A x cols_B
    C = create_matrix(rows_A, cols_B)
    for i in range(rows_A):
        for j in range(cols_B):
            for k in range(cols_A):
                C[i][j] += A[i][k] * B[k][j]
    
    # Nếu ma trận kết quả chỉ có 1 cột, trả về vector
    if cols_B == 1:
        return [row[0] for row in C]
    
    return C

def matrix_subtract(A, B):
    """Trừ hai ma trận A và B"""
    m = len(A)
    n = len(A[0])
    C = create_matrix(m, n)
    
    for i in range(m):
        for j in range(n):
            C[i][j] = A[i][j] - B[i][j]
    
    return C

def num_multiply(num, A):
    """Nhân ma trận A với một số"""
    m = len(A)
    n = len(A[0])
    C = create_matrix(m, n)
    
    for i in range(m):
        for j in range(n):
            C[i][j] = num * A[i][j]
    
    return C

def determinant(A):
    """Tính định thức của ma trận A bằng phương pháp khai triển theo cột"""
    n = len(A)
    
    # Trường hợp cơ bản
    if n == 1:
        return A[0][0]
    if n == 2:
        return A[0][0] * A[1][1] - A[0][1] * A[1][0]
    
    det = 0
    for j in range(n):
        # Tạo ma trận con bằng cách loại bỏ hàng 0 và cột j
        minor = create_matrix(n-1, n-1)
        for i in range(1, n):
            minor_col = 0
            for k in range(n):
                if k != j:
                    minor[i-1][minor_col] = A[i][k]
                    minor_col += 1
        
        cofactor = A[0][j] * ((-1) ** j) * determinant(minor) 
        det += cofactor
    
    return det

def round_matrix(A, decimals = 10):
    """Làm tròn các phần tử của ma trận"""
    result = []
    for row in A:
        result.append([round(x, decimals) for x in row])
    return result

def print_matrix(A):
    for row in A:
        print([round(x, 10) for x in row])

def inverse(P):
    """Tính ma trận nghịch đảo của ma trận P bằng phương pháp Gauss - Jordan"""
    n = len(P)
    
    # Tạo ma trận mở rộng [P | I]
    P_augmented = []
    for i in range(n):
        row = P[i][:]
        for j in range(n):
            if i == j:
                row.append(1.0)
            else:
                row.append(0.0)
        P_augmented.append(row)
    
    # Biến đổi Gauss - Jordan
    for i in range(n):
        # Tìm pivot
        max_val = abs(P_augmented[i][i])
        max_row = i
        for k in range(i + 1, n):
            if abs(P_augmented[k][i]) > max_val:
                max_val = abs(P_augmented[k][i])
                max_row = k
        
        # Hoán đổi hàng
        if max_row != i:
            P_augmented[i], P_augmented[max_row] = P_augmented[max_row], P_augmented[i]
        
        # Chuẩn hóa hàng pivot
        pivot = P_augmented[i][i]
        for j in range(i, 2*n):
            P_augmented[i][j] /= pivot
        
        # Khử tất cả các hàng khác
        for k in range(n):
            if k != i:
                factor = P_augmented[k][i]
                for j in range(i, 2*n):
                    P_augmented[k][j] -= factor * P_augmented[i][j]
    
    # Lấy ma trận nghịch đảo
    P_inverse = []
    for i in range(n):
        P_inverse.append(P_augmented[i][n:])
    
    return P_inverse

# Hàm để tính chuyển vị của ma trận
def transpose(A):
    # Kiểm tra nếu A là vector
    if not isinstance(A[0], list):
        return [[a] for a in A]  # Chuyển vector hàng thành vector cột
    
    # Tính ma trận chuyển vị
    rows = len(A)
    cols = len(A[0])
    A_T = [[0 for _ in range(rows)] for _ in range(cols)]
    
    for i in range(rows):
        for j in range(cols):
            A_T[j][i] = A[i][j]
    
    return A_T

In [3]:
# Các hàm tính toán vector cơ bản

def mean(values):
    """Tính giá trị trung bình của một list"""
    return sum(values) / len(values) if values else 0

def sum_of_squares(values):
    """Tính tổng bình phương của một list"""
    return sum(x * x for x in values)

def subtract_constant(values, constant):
    """Trừ một hằng số từ mỗi phần tử trong danh sách"""
    return [x - constant for x in values]

---
## **4. Xây dựng mô hình đánh giá chất lượng rượu**
***Sử dụng phương pháp hồi quy tuyến tính***

In [4]:
import pandas as pd
df = pd.read_csv('wine.csv')

# loại bỏ dấu ngoặc kép trong tên cột
df.columns = df.columns.str.replace('"', '')

In [5]:
df

quality  fixed acidity  volatile acidity  citric acid  residual sugar  \
0           5            7.4             0.700         0.00             1.9   
1           5            7.8             0.880         0.00             2.6   
2           5            7.8             0.760         0.04             2.3   
3           6           11.2             0.280         0.56             1.9   
4           5            7.4             0.700         0.00             1.9   
...       ...            ...               ...          ...             ...   
1194        6            7.0             0.745         0.12             1.8   
1195        6            6.2             0.430         0.22             1.8   
1196        6            7.9             0.580         0.23             2.3   
1197        6            7.7             0.570         0.21             1.5   
1198        6            7.7             0.260         0.26             2.0   

      chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0         0.076                 11.0                    34  0.99780  3.51   
1         0.098                 25.0                    67  0.99680  3.20   
2         0.092                 15.0                    54  0.99700  3.26   
3         0.075                 17.0                    60  0.99800  3.16   
4         0.076                 11.0                    34  0.99780  3.51   
...         ...                  ...                   ...      ...   ...   
1194      0.114                 15.0                    64  0.99588  3.22   
1195      0.078                 21.0                    56  0.99633  3.52   
1196      0.076                 23.0                    94  0.99686  3.21   
1197      0.069                  4.0                     9  0.99458  3.16   
1198      0.052                 19.0                    77  0.99510  3.15   

      sulphates  alcohol  
0          0.56      9.4  
1          0.68      9.8  
2          0.65      9.8  
3          0.58      9.8  
4          0.56      9.4  
...         ...      ...  
1194       0.59      9.5  
1195       0.60      9.5  
1196       0.58      9.5  
1197       0.54      9.8  
1198       0.79     10.9  

[1199 rows x 12 columns]

In [6]:
# Overview
print("Hiển thị vài dòng đầu tiên của tập dữ liệu:")
print(df.head())

print("Mô tả thống kê:")
print(df.describe())

print("Thông tin:")
print(df.info())

Hiển thị vài dòng đầu tiên của tập dữ liệu:
   quality  fixed acidity  volatile acidity  citric acid  residual sugar  \
0        5            7.4              0.70         0.00             1.9   
1        5            7.8              0.88         0.00             2.6   
2        5            7.8              0.76         0.04             2.3   
3        6           11.2              0.28         0.56             1.9   
4        5            7.4              0.70         0.00             1.9   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.076                 11.0                    34   0.9978  3.51   
1      0.098                 25.0                    67   0.9968  3.20   
2      0.092                 15.0                    54   0.9970  3.26   
3      0.075                 17.0                    60   0.9980  3.16   
4      0.076                 11.0                    34   0.9978  3.51   

   sulphates  alcohol  
0       0.56      9.4  
1     

### **a. Sử dụng toàn bộ 11 đặc trưng đề bài cung cấp**

***Các hàm sử dụng***

In [7]:
# Hàm triển khai OLS
def ols(X, y):
    """
    Công thức: beta = (X^T X)^(-1) X^T y
    """
     # Tính ma trận X^T
    X_T = transpose(X)
    
    # Tính (X^T X)
    X_T_X = matrix_multiply(X_T, X)
    X_T_X_inverse = inverse(X_T_X)
    
    # Tính ma trận (X^T y)
    X_T_y = matrix_multiply(X_T, y)
    
    # Tính hệ số beta
    beta = matrix_multiply(X_T_X_inverse, X_T_y)
    
    return beta

# Hàm dự đoán sử dụng mô hình OLS
def predict(X, coefficients):

    # Thêm cột hằng số 1 để tính hệ số chặn
    X_with_intercept = []
    for row in X:
        X_with_intercept.append([1] + row)
    
    # Tính giá trị dự đoán y_hat = X * beta
    y_pred = []
    for row in X_with_intercept:
        pred = 0
        for j in range(len(coefficients)):
            pred += row[j] * coefficients[j]
        y_pred.append(pred)
    
    return y_pred

def evaluate_model(y_true, y_pred):
    """
    Tính các chỉ số đánh giá mô hình: MSE, R^2, chuẩn vector của phần dư
    """
    # Tính Mean Squared Error (MSE)
    squared_errors = [(y_true[i] - y_pred[i]) ** 2 for i in range(len(y_true))]
    mse = sum(squared_errors) / len(y_true)
    
    # Tính giá trị trung bình của y_true
    y_mean = mean(y_true)
    
    # Tính Total Sum of Squares (TSS)
    tss = sum_of_squares(subtract_constant(y_true, y_mean))
    
    # Tính Residual Sum of Squares (RSS)
    rss = sum(squared_errors)
    
    # Tính hệ số xác định R^2
    r_squared = 1 - (rss / tss)

    # Tính các phần dư
    residuals = [y_true[i] - y_pred[i] for i in range(len(y_true))]

    # Tính chuẩn vector của phần dư
    residuals_norm = sqrt(sum_of_squares(residuals))
    
    return mse, r_squared, residuals_norm

***Chương trình chính***

Nhắc lại mô hình tuyến tính theo tham số (linear in the parameters model) cho lớp hàm dự báo:

$$\hat{f}(x) = \theta_1 f_1(x) + \cdots + \theta_p f_p(x)$$

Trong đó:
- $f_1, f_2, ..., f_p$: $R^n \rightarrow R$ là các hàm cơ sở (basis functions) được xác định trước
- $\theta_1, \theta_2, ..., \theta_p$ là các tham số mô hình (model parameters) cần tìm

Ở đây ta chọn các hàm cơ sở chính là các đặc trưng (tính chất) của rượu mà đề bài cung cấp, tức là mỗi $f_i$ tương ứng với một đặc trưng.

Các hệ số $\beta$ chính là các tham số mô hình, trong đó có $\beta_0$ là hệ số chặn (intercept).

$$\text{quality} = \beta_0 + \beta_1 \cdot \text{feature}1 + \beta_2 \cdot \text{feature}2 + ... + \beta{11} \cdot \text{feature}{11}$$

In [ ]:
print("a. Sử dụng toàn bộ đặc trưng để dự đoán chất lượng rượu.")

# Lấy ra list các cột trong DataFrame
df_columns = df.columns.tolist()

quality_col = df_columns[0]  # Cột đầu tiên là chất lượng
features = df_columns[1:]  # Các cột còn lại là đặc trưng
print(f"\nSử dụng cột '{quality_col}' làm biến phụ thuộc và các cột còn lại làm biến độc lập.")

# Tách dữ liệu thành features (X) và target (y)
X = df[features].values.tolist()
y = df[quality_col].values.tolist()

# Hiển thị thông tin cơ bản
print(f"\nSố lượng mẫu: {len(X) - 1}")
print(f"Số lượng đặc trưng: {len(X[0])}")

# Thêm cột hằng số 1 để tính hệ số chặn (intercept)
X_with_intercept = [[1] + row for row in X]

# Tính các hệ số của mô hình
coefficients = ols(X_with_intercept, y)

# Tính giá trị dự đoán
y_pred = predict(X, coefficients)

# Đánh giá mô hình
mse, r_squared, residuals_norm = evaluate_model(y, y_pred)

# In kết quả
print("\nHệ số của mô hình OLS:")
print(f"\tHệ số chặn (Intercept): {coefficients[0]:.5f}")
for i, name in enumerate(features):
    print(f"\t{name}: {coefficients[i+1]:.5f}")

# In các chỉ số đánh giá mô hình
print("\nĐánh giá mô hình:")
print(f"\tMean Squared Error (MSE): {mse:.5f}")
print(f"\tHệ số xác định (R^2): {r_squared:.5f}")
print(f"\tChuẩn vector của phần dư: {residuals_norm:.5f}")

# Phương trình hồi quy
print("\nPhương trình hồi quy:")
equation = f"quality = {coefficients[0]:.5f}"
for i, name in enumerate(features):
    coefficient = coefficients[i+1]
    sign = "+" if coefficient >= 0 else "-"
    equation += f" {sign} {abs(coefficient):.5f} * {name}"
print(equation)

a. Sử dụng toàn bộ đặc trưng để dự đoán chất lượng rượu.

Sử dụng cột 'quality' làm biến phụ thuộc và các cột còn lại làm biến độc lập.

Số lượng mẫu: 1198
Số lượng đặc trưng: 11

Hệ số của mô hình OLS:
	Hệ số chặn (Intercept): 42.91716
	fixed acidity: 0.04752
	volatile acidity: -1.06874
	citric acid: -0.26871
	residual sugar: 0.03497
	chlorides: -1.59730
	free sulfur dioxide: 0.00349
	total sulfur dioxide: -0.00380
	density: -39.46908
	pH: -0.24558
	sulphates: 0.77384
	alcohol: 0.26938

Đánh giá mô hình:
	Mean Squared Error (MSE): 0.40715
	Hệ số xác định (R^2): 0.37829
	Chuẩn vector của phần dư: 22.09472

Phương trình hồi quy:
quality = 42.91716 + 0.04752 * fixed acidity - 1.06874 * volatile acidity - 0.26871 * citric acid + 0.03497 * residual sugar - 1.59730 * chlorides + 0.00349 * free sulfur dioxide - 0.00380 * total sulfur dioxide - 39.46908 * density - 0.24558 * pH + 0.77384 * sulphates + 0.26938 * alcohol


### **b. Sử dụng duy nhất 1 đặc trưng cho kết quả tốt nhất**

***Dùng phương pháp K-Fold Cross Validation***

K-Fold Cross-Validation chia dữ liệu thành K phần (gần) bằng nhau. Mô hình sẽ được đào tạo trên K-1 phần và kiểm tra trên phần còn lại. Quá trình này sẽ được lặp lại K lần với mỗi phần dữ liệu một lần làm tập kiểm tra.

***Các hàm sử dụng***

In [9]:
import random

# Hàm thực hiện k-fold cross validation
def kfold_cross_validation(n_samples, n_folds=5, random_seed=42):
    """
    Triển khai phương pháp k-fold cross validation.
    Trả về danh sách các chỉ số cho tập huấn luyện và tập kiểm tra.
    """

    # Kiểm tra điều kiện
    if n_samples <= 0:
        raise ValueError("Số lượng mẫu phải lớn hơn 0")
    if n_folds <= 1:
        raise ValueError("Số lượng fold phải lớn hơn 1")
    if n_folds > n_samples:
        raise ValueError("Số lượng fold không thể lớn hơn số lượng mẫu")

    random.seed(random_seed)
    
    # Tạo chỉ số cho tất cả các mẫu
    indices = list(range(n_samples))
    
    # Xáo trộn chỉ số
    random.shuffle(indices)
    
    # Chia thành k phần gần bằng nhau
    fold_sizes = [n_samples // n_folds] * n_folds
    
    # Phân phối phần dư
    for i in range(n_samples % n_folds):
        fold_sizes[i] += 1
    
    # Tạo chỉ số cho từng fold
    current = 0
    folds = []
    for fold_size in fold_sizes:
        start = current
        stop = current + fold_size

        # Tạo chỉ số cho tập huấn luyện và tập kiểm tra
        test_indices = indices[start:stop]
        train_indices = indices[:start] + indices[stop:]

        # Lưu các chỉ số vào danh sách
        folds.append((train_indices, test_indices))
        current = stop
        
    return folds

# Hàm lấy phần tử theo chỉ số
def get_subset_by_indices(data, indices):
    return [data[i] for i in indices]

# Hàm triển khai OLS (một đặc trưng)
def ols_single_feature(X, y):

    # Tạo ma trận X với cột hằng số 1
    X_with_intercept = [[1, x] for x in X]
    
    # Gọi hàm ols để tính hệ số
    coefficients = ols(X_with_intercept, y)
    
    # Trích xuất beta_0 và beta_1
    beta_0 = coefficients[0]
    beta_1 = coefficients[1]
    
    return beta_0, beta_1

# Hàm dự đoán
def predict_single_feature(X, beta_0, beta_1):
    return [beta_0 + beta_1 * x for x in X]

"""
Hàm tính các chỉ số đánh giá mô hình: dùng lại hàm evaluate_model đã định nghĩa ở trên
"""

def rank_features_by_cv(X_all, y, feature_names, k=10, random_seed=42):
    """
    Xếp hạng các đặc trưng theo hiệu suất Cross Validation
    """
    n_features = len(feature_names)
    feature_scores = []
    
    # Tạo folds cho cross validation
    folds = kfold_cross_validation(len(y), n_folds=k, random_seed=random_seed)
    
    # Đánh giá từng đặc trưng
    for i in range(n_features):
        # Lấy dữ liệu cho đặc trưng thứ i
        X_feature = [row[i] for row in X_all]
        
        fold_r2 = []
        fold_mse = []
        fold_residuals_norm = []
        
        # Thực hiện k-fold cross validation
        for train_idx, test_idx in folds:
            X_train = get_subset_by_indices(X_feature, train_idx)
            X_test = get_subset_by_indices(X_feature, test_idx)
            y_train = get_subset_by_indices(y, train_idx)
            y_test = get_subset_by_indices(y, test_idx)
            
            # Xây dựng mô hình và đánh giá
            beta_0, beta_1 = ols_single_feature(X_train, y_train)
            y_pred = predict_single_feature(X_test, beta_0, beta_1)
            mse, r2, residuals_norm = evaluate_model(y_test, y_pred)
            
            fold_r2.append(r2)
            fold_mse.append(mse)
            fold_residuals_norm.append(residuals_norm)
        
        # Tính trung bình các chỉ số cho từng fold
        avg_mse = mean(fold_mse)
        avg_r2 = mean(fold_r2)
        avg_residuals_norm = mean(fold_residuals_norm)
        feature_scores.append((feature_names[i], avg_mse, avg_r2, avg_residuals_norm))
    
    # Xếp hạng theo R^2 giảm dần
    feature_scores.sort(key=lambda x: x[2], reverse=True)
    
    # trả về list các đặc trưng và điểm số
    return feature_scores

In [10]:
import pandas as pd

def sort_dataframe(df, column, ascending=True):
    """Sắp xếp DataFrame theo một cột cụ thể"""
    sorted_indices = sorted(range(len(df)), key=lambda i: df.iloc[i][column], reverse=not ascending)
    return df.iloc[sorted_indices].reset_index(drop=True)

# Hàm tạo bảng so sánh giữa các đặc trưng
def show_comparison_table(feature_names, all_mse, all_r2, all_residuals_norm):

    # Tạo bảng kết quả
    results = pd.DataFrame({
        'Feature': feature_names,
        'MSE': all_mse,
        'R^2': all_r2,
        'Residuals Norm': all_residuals_norm
    })    

    # Sắp xếp theo R^2 giảm dần, MSE tăng dần, và chuẩn vector phần dư tăng dần
    results_by_r2 = sort_dataframe(results, 'R^2', ascending=False)
    results_by_mse = sort_dataframe(results, 'MSE', ascending=True)
    results_by_residuals = sort_dataframe(results, 'Residuals Norm', ascending=True)

    print("\nXếp hạng theo R^2:")
    print(results_by_r2)

    print("\nXếp hạng theo MSE:")
    print(results_by_mse)

    print("\nXếp hạng theo chuẩn vector phần dư:")
    print(results_by_residuals)

***Chương trình chính***

In [11]:
print("b. Sử dụng duy nhất một đặc trưng cho kết quả tốt nhất.")

quality_col = df.columns[0]  # Cột chất lượng
feature_columns = df.columns[1:].values.tolist()  # Các cột còn lại là đặc trưng

# Danh sách các chỉ số đánh giá
all_mse = []  
all_r2 = []
all_residuals_norm = []
feature_names = []

X_all = df[feature_columns].values.tolist()
y = df[quality_col].values.tolist()  # Chất lượng rượu

feature_rankings = rank_features_by_cv(X_all, y, feature_columns, k=10, random_seed=42)

for feature, mse, r2, residuals_norm in feature_rankings:
    all_mse.append(mse)
    all_r2.append(r2)
    all_residuals_norm.append(residuals_norm)
    feature_names.append(feature)

# In bảng kết quả
show_comparison_table(feature_names, all_mse, all_r2, all_residuals_norm)

# Tìm đặc trưng tốt nhất
best_feature_r2_idx = all_r2.index(max(all_r2))
best_feature_r2 = feature_names[best_feature_r2_idx]

best_feature_mse_idx = all_mse.index(min(all_mse))
best_feature_mse = feature_names[best_feature_mse_idx]

best_feature_resnorm_idx = all_residuals_norm.index(min(all_residuals_norm))
best_feature_resnorm = feature_names[best_feature_resnorm_idx]

print(f"\nĐặc trưng tốt nhất theo R^2: {best_feature_r2}")
print(f"Đặc trưng tốt nhất theo MSE: {best_feature_mse}")
print(f"Đặc trưng tốt nhất theo chuẩn vector phần dư: {best_feature_resnorm}")

# Phân tích chi tiết đặc trưng tốt nhất
best_feature_col = best_feature_r2.strip('"')
X_best = df[best_feature_col].values.tolist()
y = df[quality_col].values.tolist()

# Tính hệ số cho toàn bộ dữ liệu
beta_0, beta_1 = ols_single_feature(X_best, y)
y_pred = predict_single_feature(X_best, beta_0, beta_1)
mse, r2, residuals_norm = evaluate_model(y, y_pred)

print(f"\nMô hình cho đặc trưng tốt nhất ({best_feature_r2}):")
print(f"\tHệ số chặn (beta_0): {beta_0:.5f}")
print(f"\tHệ số góc (beta_1): {beta_1:.5f}")
print(f"\tMSE: {mse:.5f}")
print(f"\tR^2: {r2:.5f}")
print(f"\tChuẩn vector của phần dư: {residuals_norm:.5f}")
print(f"\nPhương trình hồi quy:")
print(f"quality = {beta_0:.5f} + {beta_1:.5f} * {best_feature_r2}")


b. Sử dụng duy nhất một đặc trưng cho kết quả tốt nhất.

Xếp hạng theo R^2:
                 Feature       MSE       R^2  Residuals Norm
0                alcohol  0.488588  0.249910        7.633226
1       volatile acidity  0.562125  0.135573        8.192508
2   total sulfur dioxide  0.622987  0.045722        8.620883
3            citric acid  0.623031  0.045071        8.621715
4              sulphates  0.626997  0.039782        8.649428
5                density  0.632577  0.031890        8.685438
6          fixed acidity  0.641988  0.015806        8.754422
7              chlorides  0.646725  0.008614        8.786693
8                     pH  0.652735 -0.000419        8.827413
9    free sulfur dioxide  0.654005 -0.002268        8.835621
10        residual sugar  0.655485 -0.004708        8.846259

Xếp hạng theo MSE:
                 Feature       MSE       R^2  Residuals Norm
0                alcohol  0.488588  0.249910        7.633226
1       volatile acidity  0.562125  0.135573      

Việc chọn K = 5 hoặc K = 10 là những lựa chọn phổ biến vì chúng mang lại sự cân bằng tốt giữa hiệu quả tính toán và khả năng ước lượng hiệu suất của mô hình. Ở đây với bộ dữ liệu (dataset) gồm 1198 dòng, K = 10 là một lựa chọn phù hợp, cân bằng tốt giữa độ tin cậy của đánh giá và hiệu quả tính toán.

### **c. Xây dựng mô hình của riêng tôi cho kết quả tốt nhất**

***Các hàm quan trọng***

In [12]:
import pandas as pd
import random

# Hàm để tách dữ liệu thành tập huấn luyện và tập kiểm tra
def train_test_split(X, y, test_ratio=0.2, random_seed=42):
    """
    Chia dữ liệu thành tập huấn luyện và tập kiểm tra.
    
    Tham số:
    X: Ma trận các đặc trưng
    y: Vector giá trị mục tiêu (quality)
    test_ratio: Tỷ lệ dữ liệu dùng cho tập kiểm tra
    random_seed: Giá trị seed để tái tạo kết quả
    """

    random.seed(random_seed)
    n_samples = len(X)
    indices = list(range(n_samples))
    random.shuffle(indices)
    
    test_size = int(n_samples * test_ratio)
    test_indices = indices[:test_size]
    train_indices = indices[test_size:]
    
    X_train = get_subset_by_indices(X, train_indices)
    X_test = get_subset_by_indices(X, test_indices)
    y_train = get_subset_by_indices(y, train_indices)
    y_test = get_subset_by_indices(y, test_indices)
    
    return X_train, X_test, y_train, y_test

# Hàm tính toán độ lệch chuẩn
def std_dev(values):
    mean_value = mean(values)
    variance = sum((x - mean_value) ** 2 for x in values) / len(values)
    return sqrt(variance)

# Hàm chuẩn hóa dữ liệu theo phương pháp Z-score
def standardize(X):
    # Nếu X là vector
    if not isinstance(X[0], list):
        n_samples = len(X)
        
        # Tính mean và std
        mean_value = mean(X)
        std_value = std_dev(X)
        
        # Chuẩn hóa
        X_standardized = [(x - mean_value) / (std_value if std_value > 0 else 1) for x in X]
        
        return X_standardized
    
    # Nếu X là matrix
    n_features = len(X[0])
    n_samples = len(X)
    
    # Tính mean và std cho mỗi feature
    means = []
    stds = []
    
    for j in range(n_features):
        feature_values = [X[i][j] for i in range(n_samples)]
        means.append(mean(feature_values))
        stds.append(std_dev(feature_values))
    
    # Chuẩn hóa
    X_standardized = []
    for i in range(n_samples):
        row = []
        for j in range(n_features):
            row.append((X[i][j] - means[j]) / stds[j])
        X_standardized.append(row)
    
    return X_standardized

***Chương trình chính***

In [13]:
print("c. Xây dựng mô hình tối ưu cho dự đoán chất lượng rượu")

# 1. Dùng cross validation để đánh giá các đặc trưng
feature_columns = df.columns[1:].values.tolist()
quality_col = df.columns[0]

X_all = df[feature_columns].values.tolist()
y = df[quality_col].values.tolist()

feature_ranking = rank_features_by_cv(X_all, y, feature_columns, k=10, random_seed=42)

feature_names, all_mse, all_r2, all_residuals_norm = zip(*feature_ranking)
print("\nXếp hạng các đặc trưng theo hiệu suất Cross Validation:")
show_comparison_table(feature_names, all_mse, all_r2, all_residuals_norm)

# 2. Chọn các đặc trưng tốt nhất dựa trên cross validation
feature_counts = list(range(1, len(feature_columns) + 1))
r2_scores = []
mse_scores = []
residuals_norm_scores = []

print("\nThử nghiệm với các số lượng đặc trưng khác nhau:")
print("-" * 50)

for n_features in feature_counts:
    # Lấy n đặc trưng quan trọng nhất
    top_n_features = [feature[0] for feature in feature_ranking[:n_features]]

    # 3. Lấy dữ liệu cho các đặc trưng đã chọn
    feature_indices = [feature_columns.index(feature) for feature in top_n_features]
    X_selected = [[row[i] for i in feature_indices] for row in X_all]
    
    # 4. Chuẩn hóa dữ liệu
    X_standardized = standardize(X_selected)
    
    # 5. Chia dữ liệu thành tập huấn luyện và tập kiểm tra
    X_train, X_test, y_train, y_test = train_test_split(X_standardized, y, test_ratio=0.2, random_seed=42)
    
    # 6. Thêm cột hằng số 1 cho hệ số chặn
    X_train_intercept = [[1] + row for row in X_train]
    
    # 7. Huấn luyện mô hình
    coefficients = ols(X_train_intercept, y_train)
    
    # 8. Dự đoán và đánh giá trên tập kiểm tra
    y_pred_test = predict(X_test, coefficients)
    mse, r2, residuals_norm = evaluate_model(y_test, y_pred_test)
    
    # Lưu kết quả
    r2_scores.append(r2)
    mse_scores.append(mse)
    residuals_norm_scores.append(residuals_norm)
    
    print(f"Số đặc trưng: {n_features}, R^2: {r2:.5f}, MSE: {mse:.5f}, Res_norms: {residuals_norm:.5f}")

# Tìm số đặc trưng tối ưu
optimal_n_features = feature_counts[r2_scores.index(max(r2_scores))]
print(f"\nSố đặc trưng tối ưu (theo R^2): {optimal_n_features}")

# Sử dụng số đặc trưng tối ưu tìm được
print(f"Xây dựng mô hình tối ưu với {optimal_n_features} đặc trưng")
top_features = [feature[0] for feature in feature_ranking[:optimal_n_features]]

mse_test = mse_scores[optimal_n_features - 1]
r2_test = r2_scores[optimal_n_features - 1]
residuals_norm = residuals_norm_scores[optimal_n_features - 1]

# In kết quả
print("\nKết quả mô hình tối ưu (trên tập kiểm tra):")
print(f"\tMSE: {mse_test:.5f}")
print(f"\tR^2: {r2_test:.5f}")
print(f"\tChuẩn vector phần dư: {residuals_norm:.5f}")

# In hệ số mô hình
print("\nHệ số của mô hình:")
print(f"\tHệ số chặn (beta_0): {coefficients[0]:.5f}")
for i, feature in enumerate(top_features):
    print(f"\t{feature}: {coefficients[i+1]:.5f}")

# Phương trình hồi quy
print("\nPhương trình hồi quy:")
equation = f"quality = {coefficients[0]:.5f}"
for i, feature in enumerate(top_features):
    coefficient = coefficients[i+1]
    sign = "+" if coefficient >= 0 else "-"
    equation += f" {sign} {abs(coefficient):.5f} * {feature}"
print(equation)

c. Xây dựng mô hình tối ưu cho dự đoán chất lượng rượu

Xếp hạng các đặc trưng theo hiệu suất Cross Validation:

Xếp hạng theo R^2:
                 Feature       MSE       R^2  Residuals Norm
0                alcohol  0.488588  0.249910        7.633226
1       volatile acidity  0.562125  0.135573        8.192508
2   total sulfur dioxide  0.622987  0.045722        8.620883
3            citric acid  0.623031  0.045071        8.621715
4              sulphates  0.626997  0.039782        8.649428
5                density  0.632577  0.031890        8.685438
6          fixed acidity  0.641988  0.015806        8.754422
7              chlorides  0.646725  0.008614        8.786693
8                     pH  0.652735 -0.000419        8.827413
9    free sulfur dioxide  0.654005 -0.002268        8.835621
10        residual sugar  0.655485 -0.004708        8.846259

Xếp hạng theo MSE:
                 Feature       MSE       R^2  Residuals Norm
0                alcohol  0.488588  0.249910        7.6

**So sánh các mô hình**

In [14]:
# Đánh giá mô hình sử dụng tất cả đặc trưng
X_all_standardized = standardize(X_all)     # Chuẩn hóa tất cả đặc trưng

X_all_train, X_all_test, y_all_train, y_all_test = train_test_split(X_all_standardized, y, test_ratio=0.2, random_seed=42)

X_all_train_intercept = [[1] + row for row in X_all_train]
coefficients_all = ols(X_all_train_intercept, y_all_train)
y_pred_all_test = predict(X_all_test, coefficients_all)
mse_all_test, r2_all_test, res_norm_all_test = evaluate_model(y_all_test, y_pred_all_test)


# Đánh giá mô hình đơn đặc trưng tốt nhất (alcohol)
best_feature = feature_ranking[0][0]
best_feature_index = feature_columns.index(best_feature)

# Lấy dữ liệu cho đặc trưng tốt nhất
X_best = [row[best_feature_index] for row in X_all]
X_best_standardized = standardize(X_best)   # Chuẩn hóa đặc trưng tốt nhất

X_best_train, X_best_test, y_best_train, y_best_test = train_test_split(X_best_standardized, y, test_ratio=0.2, random_seed=42)
beta_0, beta_1 = ols_single_feature(X_best_train, y_best_train)
y_pred_best_test = predict_single_feature(X_best_test, beta_0, beta_1)
mse_best_test, r2_best_test, res_norm_best_test = evaluate_model(y_best_test, y_pred_best_test)


# Tạo DataFrame để so sánh các mô hình
comparison_data = {
    'Mô hình': [f'Tối ưu ({optimal_n_features} đặc trưng)', 'Tất cả (11 đặc trưng)', f'Đơn đặc trưng ({best_feature})'],
    'MSE': [mse_test, mse_all_test, mse_best_test],
    'R^2': [r2_test, r2_all_test, r2_best_test],
    'Chuẩn vector phần dư': [residuals_norm, res_norm_all_test, res_norm_best_test]
}

comparison_df = pd.DataFrame(comparison_data)
comparison_df.index = range(1, len(comparison_df) + 1)

print("\nSo sánh các mô hình (đã chuẩn hóa dữ liệu):")
print(comparison_df)

# Phân tích kết quả
best_model_idx = comparison_df['R^2'].map(lambda x: float(x)).idxmax()
best_model = comparison_df.iloc[best_model_idx - 1]['Mô hình']
best_r2 = comparison_df.iloc[best_model_idx - 1]['R^2']

print(f"\nMô hình tốt nhất là: {best_model} với R^2 = {best_r2}")


So sánh các mô hình (đã chuẩn hóa dữ liệu):
                   Mô hình       MSE       R^2  Chuẩn vector phần dư
1     Tối ưu (8 đặc trưng)  0.375494  0.371120               9.47328
2    Tất cả (11 đặc trưng)  0.394365  0.339514               9.70841
3  Đơn đặc trưng (alcohol)  0.441106  0.261233              10.26763

Mô hình tốt nhất là: Tối ưu (8 đặc trưng) với R^2 = 0.37112040766195264


***
## **5. Mô tả ý tưởng và các hàm**
### **Ý tưởng**

> Chương trình thực hiện bài toán Data fitting (Khớp dữ liệu) sử dụng phương pháp hồi quy tuyến tính (Linear Regression) bằng thuật toán Bình phương tối thiểu OLS (Ordinary Least Squares) để đánh giá chất lượng rượu dựa trên 11 đặc trưng khác nhau.

Chương trình thực hiện 3 phương pháp tiếp cận:
1. **Sử dụng toàn bộ 11 đặc trưng**
   - Sử dụng tất cả đặc trưng để xây dựng mô hình hồi quy tuyến tính
   - Đánh giá hiệu suất mô hình thông qua $R^2$, MSE và chuẩn phần dư

2. **Tìm đặc trưng đơn lẻ tốt nhất thông qua K-Fold Cross Validation**
   - Chia dữ liệu thành k = 10 fold để đảm bảo kết quả đánh giá ổn định và đáng tin cậy
   - Xếp hạng các đặc trưng dựa trên hiệu suất trung bình qua tất cả các fold
   - Phân tích và mô hình hóa đặc trưng tốt nhất

3. **Xây dựng mô hình tối ưu:**
   - Phân tích đặc trưng quan trọng thông qua Cross Validation
   - Thử nghiệm tăng dần số lượng đặc trưng (từ 1 đến 11) để tìm số đặc trưng cho kết quả tối ưu
   - Chuẩn hóa dữ liệu bằng z-score để cải thiện hiệu suất mô hình
   - Phân tách dữ liệu thành tập huấn luyện (80%) và tập kiểm tra (20%)
   - So sánh với các mô hình khác để đánh giá hiệu quả tổng thể


### **Các hàm tiện ích (Utils)**

#### *Tính toán trên số*
- `sqrt(x)`: Tính căn bậc hai của một số

#### *Ma trận*
- `create_matrix(rows, cols, default_value=0)`: Tạo ma trận với kích thước và giá trị mặc định
- `create_identity_matrix(n)`: Tạo ma trận đơn vị kích thước n×n
- `matrix_copy(A)`: Tạo bản sao của ma trận A
- `matrix_multiply(A, B)`: Nhân hai ma trận (hỗ trợ cả ma trận và vector)
- `matrix_subtract(A, B)`: Trừ hai ma trận
- `num_multiply(num, A)`: Nhân ma trận với một số
- `transpose(A)`: Tính ma trận chuyển vị
- `determinant(A)`: Tính định thức ma trận
- `inverse(P)`: Tính ma trận nghịch đảo bằng phương pháp Gauss-Jordan
- `round_matrix(A, decimals=10)`: Làm tròn các phần tử ma trận
- `print_matrix(A)`: In ma trận

#### *Vector*
- `mean(values)`: Tính giá trị trung bình
- `sum_of_squares(values)`: Tính tổng bình phương
- `subtract_constant(values, constant)`: Trừ một hằng số từ mỗi phần tử trong danh sách
- `sqrt(x)`: Hàm tính căn bậc hai của một số

### **Các hàm hồi quy tuyến tính**

#### *Phương pháp toàn bộ đặc trưng*
- `ols(X, y)`: Thực hiện hồi quy tuyến tính bằng phương pháp bình phương tối thiểu (OLS)
  - Công thức: $$\beta = (X^T X)^{-1} X^T y$$
- `predict(X, coefficients)`: Dự đoán giá trị đầu ra dựa trên mô hình
- `evaluate_model(y_true, y_pred)`: Tính các chỉ số đánh giá MSE, R² và chuẩn vector phần dư

#### *Phương pháp đơn đặc trưng và Cross Validation*
- `kfold_cross_validation(n_samples, n_folds, random_seed)`: Chia dữ liệu thành k phần cho cross validation
- `get_subset_by_indices(data, indices)`: Lấy phần tử theo chỉ số từ danh sách
- `ols_single_feature(X, y)`: Tính hệ số hồi quy cho một đặc trưng bằng cách sử dụng hàm OLS tổng quát
- `predict_single_feature(X, beta_0, beta_1)`: Dự đoán giá trị đầu ra cho một đặc trưng
- `rank_features_by_cv(X_all, y, feature_names, k, random_seed)`: Xếp hạng đặc trưng dựa trên k-fold CV

#### *Các hàm phân tích và hiển thị*
- `sort_dataframe(df, column, ascending)`: Sắp xếp DataFrame theo một cột cụ thể
- `show_comparison_table(feature_names, all_mse, all_r2, all_residuals_norm)`: Hiển thị bảng kết quả

#### *Mô hình tối ưu*
- `train_test_split(X, y, test_ratio, random_seed)`: Chia dữ liệu thành tập huấn luyện và kiểm tra
- `std_dev(values)`: Tính độ lệch chuẩn của một danh sách giá trị
- `standardize(X)`: Chuẩn hóa dữ liệu theo phương pháp z-score (hỗ trợ cả vector và ma trận)

### **Kết quả và đánh giá**

Phương pháp xây dựng mô hình tối ưu dựa trên việc kết hợp nhiều kỹ thuật để đạt hiệu suất cao nhất:

1. **Đánh giá toàn diện đặc trưng**:
   - Sử dụng k-fold cross validation (k=10) để đánh giá độ quan trọng của từng đặc trưng
   - Xếp hạng đặc trưng theo các chỉ số R², MSE và chuẩn phần dư

2. **Tìm số lượng đặc trưng tối ưu**:
   - Tiến hành thử nghiệm thực tế với từ 1 đến 11 đặc trưng
   - Lựa chọn số đặc trưng cho $R^2$ cao nhất trên tập kiểm tra

3. **Tiền xử lý dữ liệu hiệu quả**:
   - Chuẩn hóa đặc trưng để đưa về cùng thang đo
   - Phân chia dữ liệu thành tập huấn luyện và tập kiểm tra với tỷ lệ 80/20

4. **So sánh các mô hình**:
   - So sánh ba phương pháp: mô hình tối ưu, mô hình đầy đủ và mô hình đơn đặc trưng
   - Đánh giá khách quan dựa trên tập kiểm tra chung

Kết quả cho thấy mô hình tối ưu với số lượng đặc trưng được chọn dựa trên thực nghiệm đạt hiệu suất cao hơn cả mô hình đơn đặc trưng và mô hình sử dụng tất cả đặc trưng, chứng minh tính hiệu quả của phương pháp lựa chọn đặc trưng thông minh kết hợp với kỹ thuật chuẩn hóa dữ liệu.